src/processor/snapshot/database.rs

```diff
@@ -0,0 +1,139 @@
+use std::{convert::TryFrom, path::PathBuf};
+
+use ethers::types::H256;
+use eyre::Result;
+use rocksdb::{Options, DB};
+use thiserror::Error;
+
+use super::types::{SnapshotFactoryDependency, SnapshotStorageLog};
+
+const STORAGE_LOGS: &str = "storage_logs";
+const INDEX_TO_KEY_MAP: &str = "index_to_key_map";
+const FACTORY_DEPS: &str = "factory_deps";
+const METADATA: &str = "metadata";
+
+const LAST_REPEATED_KEY_INDEX: &str = "LAST_REPEATED_KEY_INDEX";
+
+#[allow(clippy::enum_variant_names)]
+#[derive(Error, Debug)]
+pub enum DatabaseError {
+    #[error("no such key")]
+    NoSuchKey,
+}
+
+pub struct SnapshotDB {
+    db: DB,
+}
+
+impl SnapshotDB {
+    pub fn new(db_path: PathBuf) -> Result<Self> {
+        let mut db_opts = Options::default();
+        db_opts.create_missing_column_families(true);
+        db_opts.create_if_missing(true);
+
+        let db = DB::open_cf(
+            &db_opts,
+            db_path,
+            vec![METADATA, STORAGE_LOGS, INDEX_TO_KEY_MAP, FACTORY_DEPS],
+        )?;
+
+        Ok(Self { db })
+    }
+
+    pub fn get_last_repeated_key_index(&self) -> Result<u64> {
+        let metadata = self.db.cf_handle(METADATA).unwrap();
+        Ok(match self.db.get_cf(metadata, LAST_REPEATED_KEY_INDEX)? {
+            Some(idx_bytes) => u64::from_be_bytes([
+                idx_bytes[0],
+                idx_bytes[1],
+                idx_bytes[2],
+                idx_bytes[3],
+                idx_bytes[4],
+                idx_bytes[5],
+                idx_bytes[6],
+                idx_bytes[7],
+            ]),
+            None => {
+                self.db
+                    .put_cf(metadata, LAST_REPEATED_KEY_INDEX, u64::to_be_bytes(1))?;
+                0
+            }
```

Comment:
Clippy suggests to make this an if let instead:

```suggestion
        Ok(
            if let Some(idx_bytes) = self.db.get_cf(metadata, LAST_REPEATED_KEY_INDEX)? {
                u64::from_be_bytes([
                    idx_bytes[0],
                    idx_bytes[1],
                    idx_bytes[2],
                    idx_bytes[3],
                    idx_bytes[4],
                    idx_bytes[5],
                    idx_bytes[6],
                    idx_bytes[7],
                ])
            } else {
                self.db
                    .put_cf(metadata, LAST_REPEATED_KEY_INDEX, u64::to_be_bytes(1))?;
                0
            },
        )
```